### PR TITLE
Add support for MIDI Fighter Twister controller

### PR DESCRIPTION
**Pull request by ghost**
First of all, thanks for awesome project!

black_overhead

MIDI Fighter Twister is a MIDI controller which have 16 push-able knobs with LED.
I made a config for 2 deck use.
This config does:
- Sync Twister's knob state to Mixxx
- Sync Mixxx's state to Twister's LED
- Reset each knob by push it

I'll write wiki page later.

### CODE REVIEW

**Comment by Be-ing** [comment]
Hi, thanks for submitting this. I apologize for not responding for a long time. We require documentation on the wiki before merging controller mappings; otherwise it tends to not get written.

From the picture posted here, I suggest switching the positions of the BPM and gain knobs because it's conventional on DJ controllers and mixers to have the channel gain knobs at the top of the EQs.

Please use the `rate` ControlObject instead of `bpm`. This will manipulate the tempo fader on screen and you won't need any calculations with rateRange in the JS.

Otherwise, the code looks neat and tidy. Good job :+1:

It would be great to map this controller for Mixxx's effects too. This controller has a very similar layout to the [Allen & Heath Xone K2](https://mixxx.org/wiki/doku.php/allen_heath_xone_k2) and could be mapped similarly.  You may be able to reuse code from the Xone K2 mapping. If you're not interested in working on that, that's okay, it will still be good to have this simple 2 deck mapping.

**Comment by Be-ing** [comment]
Does this work with the default out-of-the-box MIDI settings for the MIDI Fighter Twister or would a user need to load a configuration with the MIDI Fighter settings program? It would be better to use the out-of-the-box MIDI settings because the MIDI Fighter settings program is not available on Linux. I have [asked DJTechTools](https://djtechtools.com/2016/07/28/midi-fighter-twister-5-hidden-features-you-should-know/#comment-2810423603) about configuring the settings via MIDI, but that's currently not supported.

**Comment by tock203** [comment]
Thanks for the response.
I swapped knob position of gain and BPM, made my code use `rate` instead of `bpm`, and wrote [a wiki page](https://mixxx.org/wiki/doku.php/dj_tech_tools_midi_fighter_twister).

> It would be great to map this controller for Mixxx's effects too.

Thanks for the suggestion but I almost never use effects so I leave that job for someone more motivated than me 😣

> Does this work with the default out-of-the-box MIDI settings for the MIDI Fighter Twister

Yes

> the MIDI Fighter settings program is not available on Linux

Yeah I don't have Windows so very frustrated by that 😫

**Comment by Be-ing** [comment]
Thank you for the update. I realized the volume faders and PFL buttons are not in this mapping, only the crossfader. I think it would be helpful to use the unmapped knobs for these, with pushing them being used for the PFL button.

**Comment by tock203** [comment]
That's right. I did it.

File: res/controllers/DJ TechTools MIDI Fighter Twister.midi.xml [hunk 1]
```diff
@@ -169,6 +169,26 @@
           <normal/>
         </options>
       </control>
+      <control>
+        <group>[Channel1]</group>
+        <key>volume</key>
```
Comment (Be-ing) [inline res/controllers/DJ TechTools MIDI Fighter Twister.midi.xml]: Is there no MIDI output for these? Or did you forget to add changes to the JS file to your commit?

Comment (tock203) [reply res/controllers/DJ TechTools MIDI Fighter Twister.midi.xml]: oh I forgot it https://github.com/mixxxdj/mixxx/pull/1778/commits/c219a6a1812a76e01b690e134610b948f5bb9da7

File: res/controllers/DJ TechTools-MIDI Fighter Twister-scripts.js [hunk 2]
```diff
@@ -23,6 +23,12 @@ var MidiFighterTwister = {
             engine.makeConnection("[Channel2]", "pregain", function(value) {
                 midi.sendShortMsg(cc, 0x03, linearize(value));
             }),
+            engine.makeConnection("[Channel1]", "volume", function(value) {
+                midi.sendShortMsg(cc, 0x08, linearize(value) * 2);
+            }),
+            engine.makeConnection("[Channel2]", "volume", function(value) {
+                midi.sendShortMsg(cc, 0x0B, linearize(value) * 2);
+            }),
```
Comment (Be-ing) [inline res/controllers/DJ TechTools-MIDI Fighter Twister-scripts.js]: What about the PFL button? I think the LED below the volume knob should show the state of that.

**Comment by Be-ing** [comment]
Great, thank you. Could you rebase this on the 2.1 branch so we can include it in the upcoming 2.1.5 release?

**Comment by tock203** [comment]
:+1:

**Pull request merged**